File: app/src/main/java/mb00/android/codehub/api/service/GitHubService.java
package mb00.android.codehub.api.service;

import java.util.List;

import io.reactivex.Observable;
import mb00.android.codehub.api.model.Branch;
import mb00.android.codehub.api.model.Code;
import mb00.android.codehub.api.model.CodeResult;
import mb00.android.codehub.api.model.Comment;
import mb00.android.codehub.api.model.Commit;
import mb00.android.codehub.api.model.Contributor;
import mb00.android.codehub.api.model.Gist;
import mb00.android.codehub.api.model.Issue;
import mb00.android.codehub.api.model.IssueResult;
import mb00.android.codehub.api.model.PullRequest;
import mb00.android.codehub.api.model.Pulse;
import mb00.android.codehub.api.model.Readme;
import mb00.android.codehub.api.model.Release;
import mb00.android.codehub.api.model.Repo;
import mb00.android.codehub.api.model.RepoResult;
import mb00.android.codehub.api.model.User;
import mb00.android.codehub.api.model.UserResult;
import retrofit2.http.GET;
import retrofit2.http.Header;
import retrofit2.http.Path;
import retrofit2.http.Query;

/**
 * Retrofit interface for the GitHub API
 */

public interface GitHubService {

    @GET("search/repositories")
    Observable<RepoResult> repoSearch(@Header("Authorization") String authHeader, @Query("q") String repo);

    @GET("search/users")
    Observable<UserResult> userSearch(@Header("Authorization") String authHeader, @Query("q") String user);

    @GET("search/code")
    Observable<CodeResult> codeSearch(@Header("Authorization") String authHeader, @Query("q") String code);

    @GET("search/issues")
    Observable<IssueResult> issueSearch(@Header("Authorization") String authHeader, @Query("q") String issue);


    @GET("users/{user}")
    Observable<User> getUserOverview(@Header("Authorization") String authHeader, @Path("user") String user);

    @GET("users/{user}/events")
    Observable<List<Pulse>> getUserPulse(@Header("Authorization") String authHeader, @Path("user") String user);

    @GET("users/{user}/repos")
    Observable<List<Repo>> getUserRepos(@Header("Authorization") String authHeader, @Path("user") String user);

    @GET("users/{user}/starred")
    Observable<List<Repo>> getUserStarred(@Header("Authorization") String authHeader, @Path("user") String user);

    @GET("users/{user}/gists")
    Observable<List<Gist>> getUserGists(@Header("Authorization") String authHeader, @Path("user") String user);

    @GET("users/{user}/followers")
    Observable<List<User>> getUserFollowers(@Header("Authorization") String authHeader, @Path("user") String users);

    @GET("users/{user}/following")
    Observable<List<User>> getUserFollowing(@Header("Authorization") String authHeader, @Path("user") String users);


    @GET("repos/{user}/{repo}/contents/{path}")
    Observable<List<Code>> getRepoContents(@Header("Authorization") String authHeader, @Path("user") String user, @Path("repo") String repo, @Path("path") String path);

    @GET("repos/{user}/{repo}/readme")
    Observable<Readme> getRepoReadme(@Header("Authorization") String authHeader, @Path("user") String user, @Path("repo") String repo);

    @GET("repos/{user}/{repo}/contents/{filePath}")
    Observable<Code> getRepoFile(@Header("Authorization") String authHeader, @Path("user") String user, @Path("repo") String repo, @Path("filePath") String filePath);

    @GET("repos/{user}/{repo}/license")
    Observable<Code> getRepoLicense(@Header("Authorization") String authHeader, @Path("user") String user, @Path("repo") String repo);

    @GET("repos/{user}/{repo}/issues")
    Observable<List<Issue>> getRepoIssues(@Header("Authorization") String authHeader, @Path("user") String user, @Path("repo") String repo);

    @GET("repos/{user}/{repo}/pulls")
    Observable<List<PullRequest>> getRepoPullRequests(@Header("Authorization") String authHeader, @Path("user") String user, @Path("repo") String repo);

    @GET("repos/{user}/{repo}/events")
    Observable<List<Pulse>> getRepoPulse(@Header("Authorization") String authHeader, @Path("user") String user, @Path("repo") String repo);

    @GET("repos/{user}/{repo}/commits")
    Observable<List<Commit>> getRepoCommits(@Header("Authorization") String authHeader, @Path("user") String user, @Path("repo") String repo);

    @GET("repos/{user}/{repo}/branches/{branch}")
    Observable<List<Branch>> getRepoBranch(@Header("Authorization") String authHeader, @Path("user") String user, @Path("repo") String repo, @Path("branch") String branch);

    @GET("repos/{user}/{repo}/releases")
    Observable<List<Release>> getRepoReleases(@Header("Authorization") String authHeader, @Path("user") String user, @Path("repo") String repo);

    @GET("{releaseDownloadPath}")
    Observable<Release> getRepoReleaseDownload(@Header("Authorization") String authHeader, @Path("releaseDownloadPath") String releaseDownloadPath);

    @GET("repos/{user}/{repo}/contributors")
    Observable<List<Contributor>> getRepoContributors(@Header("Authorization") String authHeader, @Path("user") String user, @Path("repo") String repo);


    @GET("gists/{gist}")
    Observable<Gist> getGistContents(@Header("Authorization") String authHeader, @Path("gist") String gist);

    @GET("gists/{gist}/comments")
    Observable<List<Comment>> getGistComments(@Header("Authorization") String authHeader, @Path("gist") String gist);

}

File: app/src/main/java/mb00/android/codehub/api/service/GitHubService.kt
package mb00.android.codehub.api.service

import io.reactivex.Single
import mb00.android.codehub.api.model.*
import retrofit2.http.*

/**
 * Retrofit interface for the GitHub API
 */

interface GitHubService {

    @Headers("Accept: application/json")
    @POST("login/oauth/access_token")
    @FormUrlEncoded
    fun getAccessToken(
            @Field("client_id") clientId: String,
            @Field("client_secret") clientSecret: String,
            @Field("code") code: String
    ): Single<AccessToken>

    @GET("search/repositories")
    fun repoSearch(@Header("Authorization") authHeader: String, @Query("q") repo: String): Single<RepoResult>

    @GET("search/users")
    fun userSearch(@Header("Authorization") authHeader: String, @Query("q") user: String): Single<UserResult>

    @GET("search/code")
    fun codeSearch(@Header("Authorization") authHeader: String, @Query("q") code: String): Single<CodeResult>

    @GET("search/issues")
    fun issueSearch(@Header("Authorization") authHeader: String, @Query("q") issue: String): Single<IssueResult>


    @GET("users/{user}")
    fun getUserOverview(@Header("Authorization") authHeader: String, @Path("user") user: String): Single<User>

    @GET("users/{user}/events")
    fun getUserPulse(@Header("Authorization") authHeader: String, @Path("user") user: String): Single<List<Pulse>>

    @GET("users/{user}/repos")
    fun getUserRepos(@Header("Authorization") authHeader: String, @Path("user") user: String): Single<List<Repo>>

    @GET("users/{user}/starred")
    fun getUserStarred(@Header("Authorization") authHeader: String, @Path("user") user: String): Single<List<Repo>>

    @GET("users/{user}/gists")
    fun getUserGists(@Header("Authorization") authHeader: String, @Path("user") user: String): Single<List<Gist>>

    @GET("users/{user}/followers")
    fun getUserFollowers(@Header("Authorization") authHeader: String, @Path("user") users: String): Single<List<User>>

    @GET("users/{user}/following")
    fun getUserFollowing(@Header("Authorization") authHeader: String, @Path("user") users: String): Single<List<User>>


    @GET("repos/{user}/{repo}/contents/{path}")
    fun getRepoContents(@Header("Authorization") authHeader: String, @Path("user") user: String, @Path("repo") repo: String, @Path("path") path: String): Single<List<Code>>

    @GET("repos/{user}/{repo}/readme")
    fun getRepoReadme(@Header("Authorization") authHeader: String, @Path("user") user: String, @Path("repo") repo: String): Single<Readme>

    @GET("repos/{user}/{repo}/contents/{filePath}")
    fun getRepoFile(@Header("Authorization") authHeader: String, @Path("user") user: String, @Path("repo") repo: String, @Path("filePath") filePath: String): Single<Code>

    @GET("repos/{user}/{repo}/license")
    fun getRepoLicense(@Header("Authorization") authHeader: String, @Path("user") user: String, @Path("repo") repo: String): Single<Code>

    @GET("repos/{user}/{repo}/issues")
    fun getRepoIssues(@Header("Authorization") authHeader: String, @Path("user") user: String, @Path("repo") repo: String): Single<List<Issue>>

    @GET("repos/{user}/{repo}/pulls")
    fun getRepoPullRequests(@Header("Authorization") authHeader: String, @Path("user") user: String, @Path("repo") repo: String): Single<List<PullRequest>>

    @GET("repos/{user}/{repo}/events")
    fun getRepoPulse(@Header("Authorization") authHeader: String, @Path("user") user: String, @Path("repo") repo: String): Single<List<Pulse>>

    @GET("repos/{user}/{repo}/commits")
    fun getRepoCommits(@Header("Authorization") authHeader: String, @Path("user") user: String, @Path("repo") repo: String): Single<List<Commit>>

    @GET("repos/{user}/{repo}/branches/{branch}")
    fun getRepoBranch(@Header("Authorization") authHeader: String, @Path("user") user: String, @Path("repo") repo: String, @Path("branch") branch: String): Single<List<Branch>>

    @GET("repos/{user}/{repo}/releases")
    fun getRepoReleases(@Header("Authorization") authHeader: String, @Path("user") user: String, @Path("repo") repo: String): Single<List<Release>>

    @GET("{releaseDownloadPath}")
    fun getRepoReleaseDownload(@Header("Authorization") authHeader: String, @Path("releaseDownloadPath") releaseDownloadPath: String): Single<Release>

    @GET("repos/{user}/{repo}/contributors")
    fun getRepoContributors(@Header("Authorization") authHeader: String, @Path("user") user: String, @Path("repo") repo: String): Single<List<Contributor>>


    @GET("gists/{gist}")
    fun getGistContents(@Header("Authorization") authHeader: String, @Path("gist") gist: String): Single<Gist>

    @GET("gists/{gist}/comments")
    fun getGistComments(@Header("Authorization") authHeader: String, @Path("gist") gist: String): Single<List<Comment>>

}
